commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Listener definitions.
+ *
+ * DESIGN RATIONALE:
+ * <ul>
+ *    <li>The current set of listener events that can be fired are related
+ *    to modifications, and deciding between create or modify, or providing
+ *    removed entry information, require reading the previous value. Hence,
+ *    the current set of listeners can only be provided for read-write maps.
+ *    In the future, if for example cache entry visited events are to be
+ *    supported, those would need to be associated with either a read or
+ *    read-write map, and hence the listeners interface would most likely
+ *    be split up.
```

Comment:
> I just don't agree that a user holding a read-write map will never be interested in write-only events.

Right, I think I understand your point now. I agree that a read-write map might just want to register a listener for any write operation. However, if it decides to listen for any write operation, it needs to adhere to WriteListeners.onWrite definition where it receives a read-only view of the entry so, for a insert/update, the value added, for remove, an empty value. This could be solved by making RWListeners extend WriteListeners. WDYT?